Glsl.meta.slang khr shader subgroup

jkwak-work opened the pull request.

ArielG-NV committed:
Partially Implement with tests, functions and built-in variables apart of GL_KHR_shader_subgroup; Partially resolves #3548

Partially Implement with tests, functions and built-in variables apart of GL_KHR_shader_subgroup; Partially resolves #3548

GL_KHR_shader_subgroup implemented based on https://github.com/KhronosGroup/GLSL/blob/main/extensions/khr/GL_KHR_shader_subgroup.txt

Implementation is broken down into seperate glsl extensions due to the ***large differences*** in implementation of each section, and functionality/testing.

GL_KHR_shader_subgroup_basic{
**Partially implemented**

Implementation:
    * All 9 built-in variables have been stubbed without proper value; implementation is still required for these system variables; related to #411.

    * Functions were reimplemented despite nearly mirrored HLSL functions due to:
        * hlsl.meta implementations targetting workgroups rather than a warp/wave/subgroup:
            * `__syncwarp` vs `__syncthreads`
            * `SubgroupMemory` vs `WorkgroupMemory`
            * etc.
        * hlsl.meta implementations target broader SPIR-V memory targets to block on:
            * ImageMemory|UniformMemory versus SPIR-V specifying barriers for ImageMemory and seperately an option for UniformMemory
        * `subgroupElect` for CUDA has a different implementation than `WaveIsFirstLane`, this is because spec states that `subgroupElect()` only returns the lowest active gl_SubgroupInvocationID; therefore we are supposed to fetch the current active mask even if some invocations are turned off by branches

Testing:
tests for the variable -- `tests/glsl/shader-subgroup-built-in-variables.slang`
    * these tests do not test functionality since not implemented yet

tests for the functions -- `tests/glsl/shader-subgroup-basic.slang`
    * concurrency is tested for using SubgroupMemory, UniformMemory through attempting to create a GPU side race condition with writing and reading memory
        * due to testing tools avaible there are no tests for ImageMemory
    * subgroupElect is tested to return invocation #0, the lowest invocation that will always run; wave size is 32, therefore #0 is always active and will always be the elected invocation.

}

GL_KHR_shader_subgroup_vote{
**Fully implemented**

Implementation:
    * 3/3 functions are using the hlsl.meta implementation

Testing:
`tests/glsl/shader-subgroup-vote.slang`
    * Testing each a positive (returns true) and negative (returns false) test case to ensure vote results are correct

}

GL_KHR_shader_subgroup_ballot{
**Partially implemented**

Implementation:
    There are 10/10 functions that are implemented:
    * 3 are using hlsl.meta implementation
    * 7 are using new implementations -- only support GLSL, SPIR-V, HLSL, CUDA
        * These implementations do not exist in hlsl.meta, so they were added
        * `subgroupInverseBallot` lacks an analog function to call; this feature was emulated:
            * in CUDA through knowing waves are 32bit and lanes are 0 indexed, this implys that `   (ballotResult >> YOUR_INVOCATION) & 1` checks if your invocation is active, for example, `(0b11001 >> 3) & 1` would mean that only invocation 5, 4, and 1 is active, 3 would mean `YOUR_INVOCATION` is the fourth invocation in the subgroup. `(0b11001>>3) & 1` would return true since your bit is toggled and evaluates to `0b11 & 0b1`
            * in HLSL through testing if the wave count is 32 or less (use the same logic as CUDA in this case); else find the index `YOUR_INVOCATION` corrisponds with where each vector has 32bits (32 waves); avoid division in the process. then run the same algorithm cuda employs.
            * `subgroupBallotBitExtract` is logically the same as `subgroupInverseBallot`
        * 5 implementations do not have a CUDA, HLSL, and CPP imlementation yet (subgroupBallotFindMSB, subgroupBallotFindLSB, subgroupBallotExclusiveBitCount, subgroupBallotInclusiveBitCount, subgroupBallotBitCount) due to being out of scope for the commit

Testing:
`tests/glsl/shader-subgroup-ballot.slang`
    * the function tests for an expected value of each ballot function; tests try inputting larger than 32 toggled bits as function parameters to ensure the implementation correctly identifies values up to a maximum of the subgroup invocation count as per extension specification (otherwise the functionality is fairly trivial to test)

}

GL_KHR_shader_subgroup_arithmetic{
**Partially implemented**

Implementation:
    * There are 21 functions to implement:
        * 14 functions are using the hlsl.meta implementation
        * 7 functions are new implementations -- only implemented for GLSL and SPIR-V
            * GLSL & SPIR-V both use their related functions, no emulation required
            * CUDA, CPP, HLSL are out of scope for the commit

Testing:
`tests/glsl/shader-subgroup-arithmetic.slang`
    * all tests silently kill the shader; outputted GLSL was checked, could not see an issue
    * these tests only check basic functionality and correctness of all functions implemented; not an exaustive test [further continued in "Other notes of worthy" at end of commit]

}

GL_KHR_shader_subgroup_shuffle{
**Partially implemented**

Implementation:
    * There are 2 functions to implement:
        * 1 function is using the existing hlsl.meta implmentation
        * 1 function is using a new implmentation (subgroupShuffleXor) -- only implmented for GLSL & SPIR-V
            * GLSL & SPIR-V both use their related functions, no emulation required

Testing:
`tests/glsl/shader-subgroup-shuffle.slang`
    * these tests only check basic functionality and correctness of all functions implemented; not an exaustive test [further continued in "Other notes of worthy" at end of commit]
    * tests fail with cpp due to `kIROp_WaveGetActiveMask` failing to be called

}

GL_KHR_shader_subgroup_shuffle_relative{
**Partially implemented**

Implementation:
    * There are 2 functions to implement:
        * all 2 functions are using a new implmentation -- only implmented for GLSL & SPIR-V
            * GLSL & SPIR-V both use their related functions, no emulation required

Testing:
`tests/glsl/shader-subgroup-shuffle-relative.slang`
    * these tests only check basic functionality and correctness of all functions implemented; not an exaustive test [further continued in "Other notes of worthy" at end of commit]

}

GL_KHR_shader_subgroup_clustered{
**Partially implemented**

Implementation:
    * There are 7 functions to implement:
        * all 7 functions are using a new implmentation -- only implmented for GLSL & SPIR-V
            * GLSL & SPIR-V both use their related functions, no emulation required

Testing:
`tests/glsl/shader-subgroup-shuffle-clustered.slang`
    * these tests only check basic functionality and correctness of all functions implemented; not an exaustive test [further continued in "Other notes of worthy" at end of commit]

}

GL_KHR_shader_subgroup_quad{
**Partially implemented**

Implementation:
    * There are 4 functions to implement:
        * all 4 functions are using hlsl.meta implmentations -- only implemented for GLSL & SPIR-V & HLSL

Testing:
`tests/glsl/shader-subgroup-shuffle-quad.slang`
    * these tests only check basic functionality and correctness of all functions implemented; not an exaustive test [further continued in "Other notes of worthy" at end of commit]

}

---------
Failing tests and why:

Note: due to system variables not being implemented largly for CUDA and CPP, these tests will fail (#3 and #4){
    tests/glsl/shader-subgroup-arithmetic.slang.3
    tests/glsl/shader-subgroup-arithmetic.slang.4
    tests/glsl/shader-subgroup-ballot.slang.4
    tests/glsl/shader-subgroup-basic.slang.3
    tests/glsl/shader-subgroup-basic.slang.4
    tests/glsl/shader-subgroup-quad.slang.3
    tests/glsl/shader-subgroup-quad.slang.4
    tests/glsl/shader-subgroup-vote.slang.3
    tests/glsl/shader-subgroup-vote.slang.4
}

Note: due to kIROp_WaveGetActiveMask not being loaded for cpp the following test will fail{
    tests/glsl/shader-subgroup-shuffle.slang.4
}

Note: due to a unknown silent error the following will fail [could not spot an error in the generated glsl and spir-v]{
    tests/glsl/shader-subgroup-arithmetic.slang.5 (vk)
    tests/glsl/shader-subgroup-arithmetic.slang.6 (vk)
}

Other notes of worthy:{

    * only a few types are checked currently in tests due to equality templates not allowing freely casting to int/uint, meaning to test types en-mass is not trivial and will most likley be completly replaced once templates can cast & check equality more freely.

    * did not implement vector types for any functions that may use them (mostly in reference to SPIR-V, since many may accept scalar or vector inputs); applicable to subgroup-shuffle, subgroup-shuffle-relative, subgroup-arithmetic, subgroup-shuffle, subgroup_clustered, subgroup_quad

    * did not implement checks for half floats

    * CUDA, CPP, HLSL implementations were largly out of scope and if not implemented, this is due to the implementation not being trivial

}

Random fixes encountered:{
    * hlsl.meta incorrectly sets `OpCapability` as `GroupNonUniformBallot` when the `OpCapability` should be `GroupNonUniformVote`; this is as per SPIR-V spec for all SPIR-V calls used in `GL_KHR_shader_subgroup_vote`: https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpGroupNonUniformAll
}

ArielG-NV committed:
Partially Implement with tests, functions and built-in variables apart of GL_KHR_shader_subgroup; Partially resolves #3548

Partially Implement with tests, functions and built-in variables apart of GL_KHR_shader_subgroup; Partially resolves #3548

GL_KHR_shader_subgroup implemented based on https://github.com/KhronosGroup/GLSL/blob/main/extensions/khr/GL_KHR_shader_subgroup.txt

GL_KHR_shader_subgroup_* & GLSL ref:
    * https://github.com/KhronosGroup/GLSL/blob/main/extensions/khr/GL_KHR_shader_subgroup.txt
    * https://www.khronos.org/blog/vulkan-subgroup-tutorial
    * https://www.khronos.org/assets/uploads/developers/library/2018-vulkan-devday/06-subgroups.pdf

HLSL ref:
    * https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-intrinsic-functions
    * https://github.com/Microsoft/DirectXShaderCompiler/wiki/Wave-Intrinsics

CUDA ref:
    * https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html

SPIR-V ref:
    * https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_memory_semantics_id

Implementation is broken down into seperate glsl extensions due to the ***large differences*** in implementation of each section, and functionality/testing.

GL_KHR_shader_subgroup_basic{
**Partially implemented**

Implementation:
    * All 9 built-in variables have been stubbed without proper value; implementation is still required for these system variables; related to #411.

    * Functions were reimplemented despite nearly mirrored HLSL functions due to:
        * hlsl.meta implementations targetting workgroups rather than a warp/wave/subgroup:
            * `__syncwarp` vs `__syncthreads`
            * `SubgroupMemory` vs `WorkgroupMemory`
            * etc.
        * hlsl.meta implementations target broader SPIR-V memory targets to block on:
            * ImageMemory|UniformMemory versus SPIR-V specifying barriers for ImageMemory and seperately an option for UniformMemory
        * `subgroupElect` for CUDA has a different implementation than `WaveIsFirstLane`, this is because spec states that `subgroupElect()` only returns the lowest active gl_SubgroupInvocationID; therefore we are supposed to fetch the current active mask even if some invocations are turned off by branches

Testing:
tests for the variable -- `tests/glsl/shader-subgroup-built-in-variables.slang`
    * these tests do not test functionality since not implemented yet

tests for the functions -- `tests/glsl/shader-subgroup-basic.slang`
    * concurrency is tested for using SubgroupMemory, UniformMemory through attempting to create a GPU side race condition with writing and reading memory
        * due to testing tools avaible there are no tests for ImageMemory
    * subgroupElect is tested to return invocation #0, the lowest invocation that will always run; wave size is 32, therefore #0 is always active and will always be the elected invocation.

}

GL_KHR_shader_subgroup_vote{
**Fully implemented**

Implementation:
    * 3/3 functions are using the hlsl.meta implementation

Testing:

`tests/glsl/shader-subgroup-vote.slang`
    * Testing each a positive (returns true) and negative (returns false) test case to ensure vote results are correct

}

GL_KHR_shader_subgroup_ballot{
**Partially implemented**

Implementation:
    There are 10/10 functions that are implemented:
    * 3 are using hlsl.meta implementation
    * 7 are using new implementations -- only support GLSL, SPIR-V, HLSL, CUDA
        * These implementations do not exist in hlsl.meta, so they were added
        * `subgroupInverseBallot` lacks an analog function to call; this feature was emulated:
            * in CUDA through knowing waves are 32bit and lanes are 0 indexed, this implys that `   (ballotResult >> YOUR_INVOCATION) & 1` checks if your invocation is active, for example, `(0b11001 >> 3) & 1` would mean that only invocation 5, 4, and 1 is active, 3 would mean `YOUR_INVOCATION` is the fourth invocation in the subgroup. `(0b11001>>3) & 1` would return true since your bit is toggled and evaluates to `0b11 & 0b1`
            * in HLSL through testing if the wave count is 32 or less (use the same logic as CUDA in this case); else find the index `YOUR_INVOCATION` corrisponds with where each vector has 32bits (32 waves); avoid division in the process. then run the same algorithm cuda employs.
            * `subgroupBallotBitExtract` is logically the same as `subgroupInverseBallot`
        * 5 implementations do not have a CUDA, HLSL, and CPP imlementation yet (subgroupBallotFindMSB, subgroupBallotFindLSB, subgroupBallotExclusiveBitCount, subgroupBallotInclusiveBitCount, subgroupBallotBitCount) due to being out of scope for the commit

Testing:
`tests/glsl/shader-subgroup-ballot.slang`
    * the function tests for an expected value of each ballot function; tests try inputting larger than 32 toggled bits as function parameters to ensure the implementation correctly identifies values up to a maximum of the subgroup invocation count as per extension specification (otherwise the functionality is fairly trivial to test)

}

GL_KHR_shader_subgroup_arithmetic{
**Partially implemented**

Implementation:
    * There are 21 functions to implement:
        * 14 functions are using the hlsl.meta implementation
        * 7 functions are new implementations -- only implemented for GLSL and SPIR-V
            * GLSL & SPIR-V both use their related functions, no emulation required
            * CUDA, CPP, HLSL are out of scope for the commit

Testing:
`tests/glsl/shader-subgroup-arithmetic.slang`
    * all tests silently kill the shader; outputted GLSL was checked, could not see an issue
    * these tests only check basic functionality and correctness of all functions implemented; [further continued in "Other notes of worthy" at end of commit]

}

GL_KHR_shader_subgroup_shuffle{
**Partially implemented**

Implementation:
    * There are 2 functions to implement:
        * 1 function is using the existing hlsl.meta implmentation
        * 1 function is using a new implmentation (subgroupShuffleXor) -- only implmented for GLSL & SPIR-V
            * GLSL & SPIR-V both use their related functions, no emulation required

Testing:
`tests/glsl/shader-subgroup-shuffle.slang`
    * these tests only check basic functionality and correctness of all functions implemented; [further continued in "Other notes of worthy" at end of commit]
    * tests fail with cpp due to `kIROp_WaveGetActiveMask` failing to be called

}

GL_KHR_shader_subgroup_shuffle_relative{
**Partially implemented**

Implementation:
    * There are 2 functions to implement:
        * all 2 functions are using a new implmentation -- only implmented for GLSL & SPIR-V
            * GLSL & SPIR-V both use their related functions, no emulation required

Testing:
`tests/glsl/shader-subgroup-shuffle-relative.slang`
    * these tests only check basic functionality and correctness of all functions implemented; [further continued in "Other notes of worthy" at end of commit]

}

GL_KHR_shader_subgroup_clustered{
**Partially implemented**

Implementation:
    * There are 7 functions to implement:
        * all 7 functions are using a new implmentation -- only implmented for GLSL & SPIR-V
            * GLSL & SPIR-V both use their related functions, no emulation required

Testing:
`tests/glsl/shader-subgroup-shuffle-clustered.slang`
    * these tests only check basic functionality and correctness of all functions implemented; [further continued in "Other notes of worthy" at end of commit]

}

GL_KHR_shader_subgroup_quad{
**Partially implemented**

Implementation:
    * There are 4 functions to implement:
        * all 4 functions are using hlsl.meta implmentations -- only implemented for GLSL & SPIR-V & HLSL

Testing:
`tests/glsl/shader-subgroup-shuffle-quad.slang`
    * these tests only check basic functionality and correctness of all functions implemented; [further continued in "Other notes of worthy" at end of commit]

}

---------
Failing tests and why:

Note: test numbers are assuming none of the existing tests are toggled off

Note: due to system variables not being implemented largly for CUDA and CPP, these tests will fail (#3 and #4){
    tests/glsl/shader-subgroup-arithmetic.slang.3
    tests/glsl/shader-subgroup-arithmetic.slang.4
    tests/glsl/shader-subgroup-ballot.slang.4
    tests/glsl/shader-subgroup-basic.slang.3
    tests/glsl/shader-subgroup-basic.slang.4
    tests/glsl/shader-subgroup-quad.slang.3
    tests/glsl/shader-subgroup-quad.slang.4
    tests/glsl/shader-subgroup-vote.slang.3
    tests/glsl/shader-subgroup-vote.slang.4
}

Note: due to kIROp_WaveGetActiveMask not being loaded for cpp the following test will fail{
    tests/glsl/shader-subgroup-shuffle.slang.4
    tests/glsl/shader-subgroup-shuffle-relative.slang.4
    tests/glsl/shader-subgroup-basic.slang.4
}

Other notes of worthy:{

    * added preamble function and macros for implementing subgroup functionality (and tests) to make it possible to iterate on the functionality with reasonable effort in the future

    * CUDA, CPP, HLSL implementations were largly out of scope and not implemented, this is due to the implementation being non trivial for many functions

    * doubles cause a silent crash on most subgroup functions tested (silent shader hang)

    * __requireGLSLExtension does not work as intended inside glsl.meta; as a result half, int16, int64 int8, all are ommited from testing

}

Random fixes encountered:{
    * hlsl.meta incorrectly sets `OpCapability` as `GroupNonUniformBallot` when the `OpCapability` should be `GroupNonUniformVote`; this is as per SPIR-V spec for all SPIR-V calls used in `GL_KHR_shader_subgroup_vote`: https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpGroupNonUniformAll
    * hlsl.meta incorrectly uses for WaveMaskPrefixBitOr (SPIR-V) OpGroupNonUniformBitwiseAnd intead of OpGroupNonUniformBitwiseOr; this was fixed
}

ArielG-NV committed:
added vector types and tests;

Partially Implement with tests, functions and built-in variables apart of GL_KHR_shader_subgroup; Partially resolves #3548

GL_KHR_shader_subgroup implemented based on https://github.com/KhronosGroup/GLSL/blob/main/extensions/khr/GL_KHR_shader_subgroup.txt

GL_KHR_shader_subgroup_* & GLSL ref:
    * https://github.com/KhronosGroup/GLSL/blob/main/extensions/khr/GL_KHR_shader_subgroup.txt
    * https://www.khronos.org/blog/vulkan-subgroup-tutorial
    * https://www.khronos.org/assets/uploads/developers/library/2018-vulkan-devday/06-subgroups.pdf

HLSL ref:
    * https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-intrinsic-functions
    * https://github.com/Microsoft/DirectXShaderCompiler/wiki/Wave-Intrinsics

CUDA ref:
    * https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html

SPIR-V ref:
    * https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_memory_semantics_id

Implementation is broken down into seperate glsl extensions due to the ***large differences*** in implementation of each section, and functionality/testing.

GL_KHR_shader_subgroup_basic{
**Partially implemented**

Implementation:
    * All 9 built-in variables have been stubbed without proper value; implementation is still required for these system variables; related to #411.

    * Functions were reimplemented despite nearly mirrored HLSL functions due to:
        * hlsl.meta implementations targetting workgroups rather than a warp/wave/subgroup:
            * `__syncwarp` vs `__syncthreads`
            * `SubgroupMemory` vs `WorkgroupMemory`
            * etc.
        * hlsl.meta implementations target broader SPIR-V memory targets to block on:
            * ImageMemory|UniformMemory versus SPIR-V specifying barriers for ImageMemory and seperately an option for UniformMemory
        * `subgroupElect` for CUDA has a different implementation than `WaveIsFirstLane`, this is because spec states that `subgroupElect()` only returns the lowest active gl_SubgroupInvocationID; therefore we are supposed to fetch the current active mask even if some invocations are turned off by branches

Testing:
tests for the variable -- `tests/glsl/shader-subgroup-built-in-variables.slang`
    * these tests do not test functionality since not implemented yet

tests for the functions -- `tests/glsl/shader-subgroup-basic.slang`
    * concurrency is tested for using SubgroupMemory, UniformMemory through attempting to create a GPU side race condition with writing and reading memory
        * due to testing tools avaible there are no tests for ImageMemory
    * subgroupElect is tested to return invocation #0, the lowest invocation that will always run; wave size is 32, therefore #0 is always active and will always be the elected invocation.

}

GL_KHR_shader_subgroup_vote{
**Fully implemented**

Implementation:
    * 3/3 functions are using the hlsl.meta implementation

Testing:

`tests/glsl/shader-subgroup-vote.slang`
    * Testing each a positive (returns true) and negative (returns false) test case to ensure vote results are correct

}

GL_KHR_shader_subgroup_ballot{
**Partially implemented**

Implementation:
    There are 10/10 functions that are implemented:
    * 3 are using hlsl.meta implementation
    * 7 are using new implementations -- only support GLSL, SPIR-V, HLSL, CUDA
        * These implementations do not exist in hlsl.meta, so they were added
        * `subgroupInverseBallot` lacks an analog function to call; this feature was emulated:
            * in CUDA through knowing waves are 32bit and lanes are 0 indexed, this implys that `   (ballotResult >> YOUR_INVOCATION) & 1` checks if your invocation is active, for example, `(0b11001 >> 3) & 1` would mean that only invocation 5, 4, and 1 is active, 3 would mean `YOUR_INVOCATION` is the fourth invocation in the subgroup. `(0b11001>>3) & 1` would return true since your bit is toggled and evaluates to `0b11 & 0b1`
            * in HLSL through testing if the wave count is 32 or less (use the same logic as CUDA in this case); else find the index `YOUR_INVOCATION` corrisponds with where each vector has 32bits (32 waves); avoid division in the process. then run the same algorithm cuda employs.
            * `subgroupBallotBitExtract` is logically the same as `subgroupInverseBallot`
        * 5 implementations do not have a CUDA, HLSL, and CPP imlementation yet (subgroupBallotFindMSB, subgroupBallotFindLSB, subgroupBallotExclusiveBitCount, subgroupBallotInclusiveBitCount, subgroupBallotBitCount) due to being out of scope for the commit

Testing:
`tests/glsl/shader-subgroup-ballot.slang`
    * the function tests for an expected value of each ballot function; tests try inputting larger than 32 toggled bits as function parameters to ensure the implementation correctly identifies values up to a maximum of the subgroup invocation count as per extension specification (otherwise the functionality is fairly trivial to test)

}

GL_KHR_shader_subgroup_arithmetic{
**Partially implemented**

Implementation:
    * There are 21 functions to implement:
        * 14 functions are using the hlsl.meta implementation
        * 7 functions are new implementations -- only implemented for GLSL and SPIR-V
            * GLSL & SPIR-V both use their related functions, no emulation required
            * CUDA, CPP, HLSL are out of scope for the commit

Testing:
`tests/glsl/shader-subgroup-arithmetic.slang`
    * all tests silently kill the shader; outputted GLSL was checked, could not see an issue
    * these tests only check basic functionality and correctness of all functions implemented; [further continued in "Other notes of worthy" at end of commit]

}

GL_KHR_shader_subgroup_shuffle{
**Partially implemented**

Implementation:
    * There are 2 functions to implement:
        * 1 function is using the existing hlsl.meta implmentation
        * 1 function is using a new implmentation (subgroupShuffleXor) -- only implmented for GLSL & SPIR-V
            * GLSL & SPIR-V both use their related functions, no emulation required

Testing:
`tests/glsl/shader-subgroup-shuffle.slang`
    * these tests only check basic functionality and correctness of all functions implemented; [further continued in "Other notes of worthy" at end of commit]
    * tests fail with cpp due to `kIROp_WaveGetActiveMask` failing to be called

}

GL_KHR_shader_subgroup_shuffle_relative{
**Partially implemented**

Implementation:
    * There are 2 functions to implement:
        * all 2 functions are using a new implmentation -- only implmented for GLSL & SPIR-V
            * GLSL & SPIR-V both use their related functions, no emulation required

Testing:
`tests/glsl/shader-subgroup-shuffle-relative.slang`
    * these tests only check basic functionality and correctness of all functions implemented; [further continued in "Other notes of worthy" at end of commit]

}

GL_KHR_shader_subgroup_clustered{
**Partially implemented**

Implementation:
    * There are 7 functions to implement:
        * all 7 functions are using a new implmentation -- only implmented for GLSL & SPIR-V
            * GLSL & SPIR-V both use their related functions, no emulation required

Testing:
`tests/glsl/shader-subgroup-shuffle-clustered.slang`
    * these tests only check basic functionality and correctness of all functions implemented; [further continued in "Other notes of worthy" at end of commit]

}

GL_KHR_shader_subgroup_quad{
**Partially implemented**

Implementation:
    * There are 4 functions to implement:
        * all 4 functions are using hlsl.meta implmentations -- only implemented for GLSL & SPIR-V & HLSL

Testing:
`tests/glsl/shader-subgroup-shuffle-quad.slang`
    * these tests only check basic functionality and correctness of all functions implemented; [further continued in "Other notes of worthy" at end of commit]

}

---------
Failing tests and why:

Note: test numbers are assuming none of the existing tests are toggled off

Note: due to system variables not being implemented largly for CUDA and CPP, these tests will fail (#3 and #4){
    tests/glsl/shader-subgroup-arithmetic.slang.3
    tests/glsl/shader-subgroup-arithmetic.slang.4
    tests/glsl/shader-subgroup-ballot.slang.4
    tests/glsl/shader-subgroup-basic.slang.3
    tests/glsl/shader-subgroup-basic.slang.4
    tests/glsl/shader-subgroup-quad.slang.3
    tests/glsl/shader-subgroup-quad.slang.4
    tests/glsl/shader-subgroup-vote.slang.3
    tests/glsl/shader-subgroup-vote.slang.4
}

Note: due to kIROp_WaveGetActiveMask not being loaded for cpp the following test will fail{
    tests/glsl/shader-subgroup-shuffle.slang.4
    tests/glsl/shader-subgroup-shuffle-relative.slang.4
    tests/glsl/shader-subgroup-basic.slang.4
}

Note: due to a unknown silent error the following will fail [could not spot an error in the generated glsl and spir-v]{
    tests/glsl/shader-subgroup-arithmetic.slang.5 (vk)
    tests/glsl/shader-subgroup-arithmetic.slang.6 (vk)
}

Other notes of worthy:{

    * only a few types are checked currently in arithmetic test; this is due to the test silently failing, meaning I can't actually test anything implemented

    * did not implement checks for half floats

    * CUDA, CPP, HLSL implementations were largly out of scope and not implemented, this is due to the implementation being non trivial for many functions

}

Random fixes encountered:{
    * hlsl.meta incorrectly sets `OpCapability` as `GroupNonUniformBallot` when the `OpCapability` should be `GroupNonUniformVote`; this is as per SPIR-V spec for all SPIR-V calls used in `GL_KHR_shader_subgroup_vote`: https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpGroupNonUniformAll
}

ArielG-NV committed:
redesign tests under suggestions that they should be smaller, more maintainable, and test the most amount of data reasonabley possible (balance with fast iterations);

optional double testing

varying parameter testing

most tests chain results now

ArielG-NV committed:
Merge branch 'glsl.meta.slang_khr_shader_subgroup' of https://github.com/ArielG-NV/slang into glsl.meta.slang_khr_shader_subgroup

ArielG-NV committed:
fix missing impl and merge conflict resolutions

ArielG-NV committed:
Merge branch 'master' of https://github.com/ArielG-NV/slang into glsl.meta.slang_khr_shader_subgroup

ArielG-NV committed:
reundant test code cleanup and organization

move tests to proper location (glsl-intrinsic)

clean up redundant code (input buffers)

ArielG-NV committed:
Merge branch 'shader-slang:master' into glsl.meta.slang_khr_shader_subgroup

ArielG-NV committed:
add missing logical operands support (and remove hlsl/cuda code reuse due to the functional differences) under all And, Or, Xor ops

redesign tests to conform to a better testing paradigm

jkwak-work commented on tests/glsl-intrinsic/shader-subgroup/shader-subgroup-arithmetic.slang:
```diff
@@ -0,0 +1,200 @@
+__generic<T : __BuiltinLogicalType>
+bool test1Logical() {
+    return true
+#if !defined(TARGET_HLSL) && !defined(TARGET_CUDA)
```
In order for this line to work, you need to modify the command-line argument for slangc at the line 5.
```
//TEST:SIMPLE(filecheck=CHECK_CUDA): -allow-glsl -stage compute -entry computeMain -target cuda -DTARGET_CUDA
```
You need to add -DTARGET_CUDA for targeting cuda.
You need to make the same change for HLSL

jkwak-work commented on tests/glsl-intrinsic/shader-subgroup/shader-subgroup-arithmetic.slang:
```diff
@@ -0,0 +1,200 @@
+        && subgroupInclusiveOr(T(1)) == T(1)
+#ifdef TEST_when_glsl_subgroupInclusiveXor_is_not_bugged
+        && subgroupInclusiveXor(T(1)) == T(0)
+#endif // #ifdef TEST_when_glsl_subgroupInclusiveXor_is_not_bugged
```
Or works but not Xor?
Aren't they almost same??

ArielG-NV replied:
SPIR-V and GLSL both produce different results; the code that runs is also very simple, so it looks like a different error than the glsl.meta code.

the result should be 1+1+1+1 for thread 3, glsl returns 1 (wrong), spir-v returns 4 (correct). It's likely a bug outside our system, but that is significantly outside domain of what I am doing.

jkwak-work commented on tests/glsl-intrinsic/shader-subgroup/shader-subgroup-arithmetic.slang:
```diff
@@ -0,0 +1,200 @@
+        //&& test1Logical<int64_t>()
+        //&& testVLogical<int64_t, 2>()
+        //&& testVLogical<int64_t, 3>()
+        //&& testVLogical<int64_t, 4>()
```
We only need to test int, uint and bool because that's what the spec requires.
        genIType subgroupAnd(genIType value);
        genUType subgroupAnd(genUType value);
        genBType subgroupAnd(genBType value);

ArielG-NV replied:
https://github.com/KhronosGroup/GLSL/blob/main/extensions/ext/GLSL_EXT_shader_subgroup_extended_types.txt

I assume from the ext that we will need to support int8 16 and 64 for all subgroup functions eventually (I added the preamble function needed inside main code to handle this, but the code does not work due to a bug with __requireGLSLExtension extension)

if you do want these removed regardless I can do this

jkwak-work replied:
If we want to keep them, we should make them all working.
Why don't they work when int and uint work?
If the situation is same in HLSL side, we wouldn't make it working for this task, which means we should remove from this test file for the submission because they will be out of the scope for this task.
The task is to bring the parity to HLSL and if HLSL cannot do it, we wouldn't do it for GLSL on this task; we will do it later.

ArielG-NV replied:
it is part of hlsl (should be as per hlsl spec); hlsl.meta handles the hlsl side if it is an int8 (doesn't need special code to handle the case*)

the GLSL side does not work because __requireGLSLExtension does not work with GLSL.meta

jkwak-work replied:
I am not sure what you mean by "the GLSL side does not work because __requireGLSLExtension does not work with GLSL.meta".
Did you mean that you cannot apply __requireGLSLExtension only to certain types such as int8_t?

jkwak-work commented on tests/glsl-intrinsic/shader-subgroup/shader-subgroup-arithmetic.slang:
```diff
@@ -0,0 +1,200 @@
+        //&& test1Arithmetic<half>()
+        //&& testVArithmetic<half, 2>()
+        //&& testVArithmetic<half, 3>()
+        //&& testVArithmetic<half, 4>()
```
Let's test only four types as the spec requires,
        genType  subgroupAdd(genType value);
        genIType subgroupAdd(genIType value);
        genUType subgroupAdd(genUType value);
        genDType subgroupAdd(genDType value);
That is float, int, uint, double and each of them has vector2,3,4 variants.

ArielG-NV replied:
also from https://github.com/KhronosGroup/GLSL/blob/main/extensions/ext/GLSL_EXT_shader_subgroup_extended_types.txt

we will need half's eventually

if you do want these removed regardless I can do this

jkwak-work replied:
Commented out lines in code can easily rot.
You can comment out certain test cases as a part of quick iterations of TDD method.
But it is not ideal to submit them.

I understand that we will need them eventually.
But we can have them when they are actually working.
We should either delete those or make them working.

jkwak-work commented on tests/glsl-intrinsic/shader-subgroup/shader-subgroup-arithmetic.slang:
```diff
@@ -0,0 +1,200 @@
+// not testing glsl or spirv due to crashing issue from some odd code emitting which looks right but fails
```
This comment doesn't seem to reflect the current state anymore.
It probably needs to be removed.

jkwak-work commented on tests/glsl-intrinsic/shader-subgroup/shader-subgroup-arithmetic.slang:
```diff
@@ -0,0 +1,200 @@
+//TEST:SIMPLE(filecheck=CHECK_HLSL): -allow-glsl -stage compute -entry computeMain -target hlsl
+//TEST:SIMPLE(filecheck=CHECK_CUDA): -allow-glsl -stage compute -entry computeMain -target cuda 
+// not testing cpp due to missing impl
+// TEST:SIMPLE(filecheck=CHECK_CPP):  -allow-glsl -stage compute -entry computeMain -target cpp
```
My personal preference of how to disabling test here is to use
```
//T-EST
```
Just a white space character works but little hard to see.

jkwak-work commented on tests/glsl-intrinsic/shader-subgroup/shader-subgroup-ballot.slang:
```diff
@@ -0,0 +1,127 @@
+    typealias gvec = vector<T, N>;
+
+    return true
+        && subgroupBroadcast(gvec(T(1)), 0) == gvec(T(1))
+        && subgroupBroadcastFirst(gvec(T(1))) == gvec(T(1))
```
This is a nitpick but I want to leave a quick comment; feel free to ignore.
The spec shows the signature of the function, subgroupBroadcast as following,
```
        genType  subgroupQuadBroadcast(genType value,  uint id);
        genIType subgroupQuadBroadcast(genIType value, uint id);
        genUType subgroupQuadBroadcast(genUType value, uint id);
        genBType subgroupQuadBroadcast(genBType value, uint id);
        genDType subgroupQuadBroadcast(genDType value, uint id);
```
You can imitate it as it is.
```
  typealias genType = vector<float,N>;
  typealias genIType = vector<int,N>;
  typealias genUType = vector<uint,N>;
  typealias genBType = vector<bool,N>;
  typealias genDType = vector<double,N>;
  return true
        && genType(1) == subgroupQuadBroadcast(genType(1),  uint(0))
        && genIType(1) == subgroupQuadBroadcast(genIType(1), uint(0))
        && genUType(1) == subgroupQuadBroadcast(genUType(1), uint(0))
        && genBType(1) == subgroupQuadBroadcast(genBType(1), uint(0))
        && genDType(1) == subgroupQuadBroadcast(genDType(1), uint(0))
```
It becomes little different but still very similar to how they appear on the document.

ArielG-NV replied:
I don't think I will adopt this for shader_subgroup specifically but I will for other ext's (I think it is a good idea todo when consistent). 

This can't be done to match almost exactly for some of the shader_subgroup functions (unless I do something like `genIType` and `genIVecType`), meaning there would be a consistency problem, possibly confusing since the code structure of the tests are already the same

jkwak-work commented on tests/glsl-intrinsic/shader-subgroup/shader-subgroup-ballot.slang:
```diff
@@ -0,0 +1,127 @@
+        //&& test1BroadcastX<int64_t>() // shader_subgroup_preamble --> __requireGLSLExtension fails
+        //&& testVBroadcastX<int64_t, 2>()
+        //&& testVBroadcastX<int64_t, 3>()
+        //&& testVBroadcastX<int64_t, 4>()
```
Same here.
Let's limit the types to what the spec document requires

ArielG-NV replied:
(same answer as other comments)

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -2825,21 +2825,1985 @@ public uint rayQueryGetIntersectionTypeEXT(rayQueryEXT q, bool committed)
+public uint gl_SubgroupLeMask;
+
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_basic)
+public uint gl_SubgroupLtMask;
```
Shouldn't they have "in" or "out" modifiers?

ArielG-NV replied:
this is not permitted syntax as per Slang, it will cause a compile error

built-ins must either be defined through their internal c++ side or inheriting from the C++ side's interface of a built-in

jkwak-work reviewed (commented):
The test cases look much much better.
I have some of minor comments.

ArielG-NV committed:
testing code style change to not use white space as a toggle for tests

jkwak-work commented on tests/glsl-intrinsic/shader-subgroup/shader-subgroup-clustered.slang:
```diff
@@ -15,33 +15,132 @@
+        //&& test1Arithmetic<double>() // hang
+        //&& testVArithmetic<double, 2>() 
+        //&& testVArithmetic<double, 3>()
+        //&& testVArithmetic<double, 4>()
```
I think we should get "double" type working for this task.
Can you debug and figure out what is going on?

ArielG-NV replied:
I will look into the issue with doubles; although the problem seems fundamental rather than due to glsl.meta 

the int8/16/64/half should all be working, and should work for HLSL already, they don't work because __requireGLSLExtension does not work with GLSL.meta (no clear error or issue) -- I know this is the case because I get the following type of errors when running tests with the type: ```glslang:  (32): error :  ' const int8_t' : required extension not requested: GL_EXT_shader_subgroup_extended_types_int8```

jkwak-work reviewed (commented):
I think we should get the five major types working.
I see that your tests are hanging for double-type.
We should debug them and fix them.

ArielG-NV committed:
provided crash reason for doubles (intel iris gpu's crash in glsl with doubles due to missing support in device caps [as per vulkan validation layer)

uncommented the `__requireGLSLExtension` code so once it is fixed int16/8/64/half wil work with subgroup not requiring future intervention

jkwak-work commented on tests/glsl-intrinsic/shader-subgroup/shader-subgroup-quad.slang:
```diff
@@ -0,0 +1,123 @@
+//TEST:SIMPLE(filecheck=CHECK_SPV):  -allow-glsl -stage compute -entry computeMain -target spirv
+//TEST:SIMPLE(filecheck=CHECK_HLSL): -allow-glsl -stage compute -entry computeMain -target hlsl -DTARGET_HLSL
+// not testing cuda due to missing impl
+//T-EST:SIMPLE(filecheck=CHECK_CUDA): -allow-glsl -stage compute -entry computeMain -target hlsl -DTARGET_CUDA 
```
shouldn't this have "-target cuda" instead of "-target hlsl"?

jkwak-work commented on tests/glsl-intrinsic/shader-subgroup/shader-subgroup-shuffle-relative.slang:
```diff
@@ -0,0 +1,111 @@
+// not testing hlsl due to missing impl
+//T-EST:SIMPLE(filecheck=CHECK_HLSL): -allow-glsl -stage compute -entry computeMain -target hlsl -DTARGET_HLSL
+// not testing cuda due to missing impl
+//T-EST:SIMPLE(filecheck=CHECK_CUDA): -allow-glsl -stage compute -entry computeMain -target hlsl -DTARGET_CUDA 
```
same here

jkwak-work commented on tests/glsl-intrinsic/shader-subgroup/shader-subgroup-shuffle.slang:
```diff
@@ -0,0 +1,116 @@
+//TEST:SIMPLE(filecheck=CHECK_SPV):  -allow-glsl -stage compute -entry computeMain -target spirv
+
+// not testing hlsl due to missing impl
+//T-EST:SIMPLE(filecheck=CHECK_HLSL): -allow-glsl -stage compute -entry computeMain -target hlsl -DTARGET_HLSL
```
Does HLSL have any similar functionality for subgroup-shuffle?

jkwak-work commented on tests/glsl-intrinsic/shader-subgroup/shader-subgroup-clustered.slang:
```diff
@@ -0,0 +1,146 @@
+
+// not testing hlsl due to missing impl
+//T-EST:SIMPLE(filecheck=CHECK_HLSL): -allow-glsl -stage compute -entry computeMain -target hlsl -DTARGET_HLSL
+// not testing cuda due to missing impl
```
Does HLSL have any similar functionality for subgroup-clustered?

jkwak-work commented on tests/glsl-intrinsic/shader-subgroup/shader-subgroup-builtin-variables.slang:
```diff
@@ -0,0 +1,49 @@
+
+    v = gl_SubgroupLeMask;
+
+    v = gl_SubgroupLtMask;
```
Let's remove the empty lines in the middles

jkwak-work commented on tests/glsl-intrinsic/shader-subgroup/shader-subgroup-builtin-variables.slang:
```diff
@@ -0,0 +1,49 @@
+{
+    uint data[];
+} outputBuffer;
+
```
This is unused in this test.
Let's remove it.

jkwak-work commented on tests/glsl-intrinsic/shader-subgroup/shader-subgroup-builtin-variables.slang:
```diff
@@ -0,0 +1,49 @@
+//TEST:SIMPLE(filecheck=CHECK_CUDA): -allow-glsl -stage compute -entry computeMain -target hlsl -DTARGET_CUDA 
+//TEST:SIMPLE(filecheck=CHECK_CPP):  -allow-glsl -stage compute -entry computeMain -target cpp
+
+//TEST(compute, vulkan):COMPARE_COMPUTE(filecheck-buffer=BUF):-vk -compute -entry computeMain -allow-glsl
+//TEST(compute, vulkan):COMPARE_COMPUTE(filecheck-buffer=BUF):-vk -compute -entry computeMain -allow-glsl -emit-spirv-directly
```
This testing doesn't do any.
Let's remove this.
It is giving us a wrong impression that the functionality testing is happening.

jkwak-work commented on tests/glsl-intrinsic/shader-subgroup/shader-subgroup-quad.slang:
```diff
@@ -0,0 +1,123 @@
+// not testing cuda due to missing impl
+//T-EST:SIMPLE(filecheck=CHECK_CUDA): -allow-glsl -stage compute -entry computeMain -target hlsl -DTARGET_CUDA 
+// not testing cpp due to missing impl
+//T-EST:SIMPLE(filecheck=CHECK_CPP):  -allow-glsl -stage compute -entry computeMain -target cpp
```
These lines seem redundant.

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -2825,21 +2825,1989 @@ public uint rayQueryGetIntersectionTypeEXT(rayQueryEXT q, bool committed)
+// GL_KHR_shader_subgroup_basic Built-in Variables 
+
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_basic)
+public uint gl_NumSubgroups;
```
You said "in" or "out" modifiers are not a correct syntax for these.
But I see them on the other variables.
```
public out float4 gl_Position : SV_Position;
public out float gl_PointSize : SV_PointSize;
public in vec4 gl_FragCoord : SV_Position;
public out float gl_FragDepth : SV_Depth;
public out int gl_FragStencilRef : SV_StencilRef;
```
Are you sure you cannot use in/out modifiers on the subgroup variables?

jkwak-work replied:
Is it hard to implement those variables properly?

ArielG-NV replied:
to implement the builtin's I would have to create new system variables since HLSL ones do not map over; in out requires the interface inheritance of a system variable 

The implementation of a system variable looks non trivial and quite unstructured in implementation. I noticed there was also an issue related to changing how defining them works at some level https://github.com/shader-slang/slang/issues/411

jkwak-work reviewed (commented):
I have some comments.

ArielG-NV committed:
fixing some vk validation layer errors (OpMemoryBarrier, Shuffle operations)

modified style of tests; removed redundancy (extra code that does nothing); fixed some incorrect run targets; added error reasons for all encountered problems (and if needed, a #define/#if toggle)

ArielG-NV commented:
all problems should be fixed (or noted down because they are largely out of scope) now

ArielG-NV commented:
I am also looking into adding the system variables of interest

jkwak-work commented:
I am worried that if the builtin variables don't work, most of subgroup shaders will not work.
Have you had a chance to run CTS and see how much passing rate went up?

ArielG-NV commented:
I did not run CTS yet, I will see if I can find a way to add the system variables first (hopefully not by adding hlsl functionality that is missing from hlsl spec), if I can't I will just run CTS for shader_subgroup.

jkwak-work commented:
I think your commit message needs a re-work, because it is too long and not human readable.
Here is what I like to propose as a starting point.
```
Implement subgroup extension for GLSL to bring parity to HLSL

Partially addresses #3548

This change implements following functions described in the KHR_shader_subgroup. The extension consists of the following extensions:
    GL_KHR_shader_subgroup_basic
    GL_KHR_shader_subgroup_vote
    GL_KHR_shader_subgroup_arithmetic
    GL_KHR_shader_subgroup_ballot
    GL_KHR_shader_subgroup_shuffle
    GL_KHR_shader_subgroup_shuffle_relative
    GL_KHR_shader_subgroup_clustered
    GL_KHR_shader_subgroup_quad

None of the built-in variables for subgroup extension are implemented, because it is non-trivial to implement; we will have to implement them as another task later.

Intel IGPU shows a hang issue on certain tests when the input parameter is double-type, which needs to be addressed later as a separate issue.
```
You can add more statements that can answer questions that the reviewers may have while reading through the code change.
It is safe to assume that the reviewers are able to understand the logic in the code, and the commit message shouldn't try to describe every details.

During the review process, if somebody ask you a question, you will need to make a decision whether to put the information to the commit message or not by assuming that the other reviewers may get the same questions.

jkwak-work commented on tests/glsl-intrinsic/shader-subgroup/shader-subgroup-builtin-variables.slang:
```diff
@@ -0,0 +1,34 @@
+
+// currently unused but required to be implemented once gl_NumSubgroups etc.. have functionality implemented
+//T-EST(compute, vulkan):COMPARE_COMPUTE(filecheck-buffer=BUF):-vk -compute -entry computeMain -allow-glsl
+//T-EST(compute, vulkan):COMPARE_COMPUTE(filecheck-buffer=BUF):-vk -compute -entry computeMain -allow-glsl -emit-spirv-directly
```
Shouldn't this work even when slang may not have implementations for the variables, when slangc generates GLSL source file, glslc should be able to properly compile them, isn't it?

ArielG-NV replied:
I can compile and run them, in an earlier message today I was asked to remove running them since it would confuse people into believing we are testing functionality which we are not really doing any of

ArielG-NV commented:
I suppose; 
```
Partially Implement GL_KHR_shader_subgroup; Partially resolves #3548

GL_KHR_shader_subgroup implemented based on https://github.com/KhronosGroup/GLSL/blob/main/extensions/khr/GL_KHR_shader_subgroup.txt

GL_KHR_shader_subgroup_* & GLSL ref: 
    * https://github.com/KhronosGroup/GLSL/blob/main/extensions/khr/GL_KHR_shader_subgroup.txt
    * https://www.khronos.org/blog/vulkan-subgroup-tutorial
    * https://www.khronos.org/assets/uploads/developers/library/2018-vulkan-devday/06-subgroups.pdf

HLSL ref: 
    * https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-intrinsic-functions
    * https://github.com/Microsoft/DirectXShaderCompiler/wiki/Wave-Intrinsics

CUDA ref: 
    * https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html

SPIR-V ref: 
    * https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_memory_semantics_id

This change implements following functions described in the KHR_shader_subgroup. The extension consists of the following extensions:
    GL_KHR_shader_subgroup_basic
    GL_KHR_shader_subgroup_vote
    GL_KHR_shader_subgroup_arithmetic
    GL_KHR_shader_subgroup_ballot
    GL_KHR_shader_subgroup_shuffle
    GL_KHR_shader_subgroup_shuffle_relative
    GL_KHR_shader_subgroup_clustered
    GL_KHR_shader_subgroup_quad


Notes of worthy{
    * added preamble function and macros for implementing subgroup functionality (and tests) to make it possible to iterate on the functionality with reasonable effort in the future

    * CUDA, CPP, HLSL implementations were largly out of scope and not implemented, this is due to the implementation being non trivial for many functions

    * doubles cause a silent crash on most subgroup functions tested (silent shader hang) on Intel Iris due to missing F64 support 

    * __requireGLSLExtension does not work as intended inside glsl.meta; as a result half, int16, int64 int8, all are ommited from testing for glsl shader_subgroup type extension support

    * built-in variables are only stubbed (NOT implemented) due to being a non trivial implementation (seperate task)
}

Random fixes{
    * hlsl.meta incorrectly sets `OpCapability` as `GroupNonUniformBallot` when the `OpCapability` should be `GroupNonUniformVote`; this is as per SPIR-V spec for all SPIR-V calls used in `GL_KHR_shader_subgroup_vote`: https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpGroupNonUniformAll
    * hlsl.meta incorrectly uses for WaveMaskPrefixBitOr (SPIR-V) OpGroupNonUniformBitwiseAnd intead of OpGroupNonUniformBitwiseOr; this was fixed
}

```
is a good starting commit message (all notes and random fixes which may standout as a problem during test); add as confusion is maybe shared?

jkwak-work commented on tests/glsl-intrinsic/shader-subgroup/shader-subgroup-quad.slang:
```diff
@@ -0,0 +1,120 @@
+// not testing cuda due to missing impl
+//T-EST:SIMPLE(filecheck=CHECK_CUDA): -allow-glsl -stage compute -entry computeMain -target cuda -DTARGET_CUDA 
+// not testing cpp due to missing impl
+//T-EST:SIMPLE(filecheck=CHECK_CPP):  -allow-glsl -stage compute -entry computeMain -target cpp
```
This test can run for HLSL but it cannot run for CUDA and CPP?
Doesn't HLSL have support for CUDA and CPP?
If it does, you can simply call them or copy and paste whatever HLSL does.

ArielG-NV replied:
the test does not exist for hlsl.meta (the code for CUDA/CPP also does not exist) 

most things have never been tested

jkwak-work reviewed (commented):
I have more comments.

ArielG-NV commented:
"I understand that we will need them eventually.
But we can have them when they are actually working.
We should either delete those or make them working."

the code should be working, but __requireGLSLExtension just does not work (at least in the way it should be)

I will ifdef the code under TEST_when_requireGLSLExtension_works since I agree that comments should be avoided, and will apply this for future tests

jkwak-work commented:
When we read the commit message, the summary of what is happening in the commit should be the first sentence.
It is different from an academic writing that the main idea goes to the bottom.

My personal preference is that you don't use those URLs because they may not last long enough and it is not directly related to the change; people who can read your change can probably be able to find those information by themselves.

ArielG-NV commented:
I agree with your points; the first sentence describes "what we are targeting", "how much was done", "issue ref" now

I still keep the khronos spec documents since these links won't degrade, and GLSL_EXT_shader_subgroup_extended_types is not a very obvious ext but is a "second part" to shader_subgroup
```
The following PR partially implements shader_subgroup extension(s) (GL_KHR_shader_subgroup & GLSL_EXT_shader_subgroup_extended_types); Fully implement all functions, stub all built-in variables; Partially resolves #3548

GL_KHR_shader_subgroup implemented based on [Khronos specification links]:
    * https://github.com/KhronosGroup/GLSL/blob/main/extensions/khr/GL_KHR_shader_subgroup.txt
    * https://github.com/KhronosGroup/GLSL/blob/main/extensions/ext/GLSL_EXT_shader_subgroup_extended_types.txt

This change implements following functions described in the KHR_shader_subgroup. The extension consists of the following extensions:
    GL_KHR_shader_subgroup_basic
    GL_KHR_shader_subgroup_vote
    GL_KHR_shader_subgroup_arithmetic
    GL_KHR_shader_subgroup_ballot
    GL_KHR_shader_subgroup_shuffle
    GL_KHR_shader_subgroup_shuffle_relative
    GL_KHR_shader_subgroup_clustered
    GL_KHR_shader_subgroup_quad


Notes of worthy{
    * added preamble function and macros for implementing subgroup functionality (and tests) to make it possible to iterate on the functionality with reasonable effort in the future

    * CUDA, CPP, HLSL implementations were largly out of scope and not implemented, this is due to the implementation being non trivial for many functions

    * doubles cause a silent crash on most subgroup functions tested (silent shader hang) on Intel Iris due to missing F64 support 

    * __requireGLSLExtension does not work as intended inside glsl.meta; as a result half, int16, int64 int8, all are ommited from testing for glsl shader_subgroup type extension support

    * built-in variables are only **stubbed** (not implemented) due to being a non trivial implementation (seperate task)
}

Random fixes{
    * hlsl.meta incorrectly sets `OpCapability` as `GroupNonUniformBallot` when the `OpCapability` should be `GroupNonUniformVote`; this is as per SPIR-V spec for all SPIR-V calls used in `GL_KHR_shader_subgroup_vote`: https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpGroupNonUniformAll
    * hlsl.meta incorrectly uses for WaveMaskPrefixBitOr (SPIR-V) OpGroupNonUniformBitwiseAnd intead of OpGroupNonUniformBitwiseOr; this was fixed
}
```

ArielG-NV committed:
remove comments of important tests inplace of #define over the broken feature of extended shader_subgroup types

ArielG-NV committed:
Merge branch 'shader-slang:master' into glsl.meta.slang_khr_shader_subgroup

ArielG-NV committed:
removed macros inside glsl.meta

removed erroneous __target_switch to directly call hlsl.meta function

added elaboration on the problem with __requireGLSLExtension

changed WaveMaskPrefixBit[or|and|xor] to support the expected type of <int> only as per `HLSL Shader Model 6.5` specs

removed "precision highp" since it does not affect tests

ArielG-NV committed:
changes some hlsl.meta functions used to be more appropriate (as per suggested)
WaveMask -> WaveActive.*
WaveMaskPrefix.* -> WavePrefix.*

remove __target_switch case's for unimplemented case's of intrinsics

fix _getLaneId() being removed from some regex used earlier

ArielG-NV committed:
fix usage of __target_intrinsic instead of __intrinsic_asm; silently would cause only arguments to be emmitted as return

changed usage of `__requireGLSLExtension` because now it causes a crash from the missing intrinsic (instead of a silent error)

ArielG-NV committed:
fix shader subgroup extended types support for GLSL and SPIR-V:
1. seperate intrinsic/__requireGLSL generating functionality of shader_subgroup_preamble into child function calls due to otherwise `__requireGLSLExtension` being ignored if the calling function of shader_subgroup_preamble calls an `__intrinsic_asm`
2. fixed HLSL.meta logic for wave operations (Add, Mul, exclusiveAdd, exclusiveMul) to no longer cast the input type T into a uint due to cost-of-op & crash.
    * Int8_t bit casted into uint32_t crashed the compiler. As per SPIR-V spec, OpGroupNonUniformI.* work on uint and int types meaning the function has no need to cast to a unit.
3. removed erroneous __target_switch for subgroupShuffle

ArielG-NV committed:
1. ignore tests gracefully
2. remove un-needed SPIRV capability specifying (with OpCapability)
3. clean up structure of  typeRequireChecks_shader_subgroup_GLSL
4. explain why HLSL/CUDA are not targeted for shader-subgroup-arithmetic.slang

ArielG-NV committed:
syntax changes + `property` declaration fix + builtin var glsl implementation + changed incorrect HLSL.meta assumptions

(#1)`property` declaration as *non member* implementation change/fix (all of the changes to `slang-lower-to-ir.cpp`)

using (#1), implemented subgroup builtin's for GLSL/SPIR-V; did not implement built'ins completly for HLSL/CUDA due to non trivial implementations. CPP has no implementation due to missing support of system values

changed some incorrect HLSL.meta subgroup implementation assumptions of type usage (bit casting 8bit->32bit, wrong capabilities causing errors)

dumping ast crash with spir-v when using builtin's fixed by adding the `builtin` spirv case (all of the changes to `slang-ast-dump.cpp`)

[ForceInline] addition to functions missing it

return instead of spirv_asm when empty blocks are used

jkwak-work committed:
Fix the build when you try to build slang-glslang (#3599)

csyonghe committed:
Refactor compiler option representations. (#3598)

* Refactor compiler option representation.

* Fix binary compatibility.

* Add a test for specifying compiler options at link time.

* Fix binary compatibility.

* Fix binary compatibility.

* Fix backward compatibility on matrix layout.

* Fix.

* Fix.

* Fix.

* Fix gfx.

* Fix gfx.

* Fix dynamic dispatch.

* Polish.

csyonghe committed:
Support link time type specialization. (#3604)

csyonghe committed:
Add wrapper type syntax for link time specialization. (#3606)

* Add wrapper type syntax for link time specialization.

* Cleanup.

csyonghe committed:
Language server robustness fix. (#3607)

* Language server robustness fix.

* Allow parameter name to be the same as its type.

* fix

* Fix test.

skallweitNV committed:
derive approximate DX12 shader model in vulkan (#3596)

Co-authored-by: Yong He <yonghe@outlook.com>

ArielG-NV committed:
Merge branch 'master' into glsl.meta.slang_khr_shader_subgroup
